### DISCUSSION TITLE
Missing Method for data classes

### DISCUSSION THREAD

**Discussion by CygnusX2112** [allenporter/pyrainbird#435] [Q&A]
Hello,

I have been experimenting with the various methods using the tools script and am finding there are potentially missing methods for data classes defined in the documentation. Is there currently a way to fetch "Forecast" or "Program" data types from the controller?

My end goal is to retrieve program schedules (start times, run times, soak duration, etc) and potentially also adjusted run times per zone for a given program. 

I've inferred that some pieces to the puzzle are still missing and perhaps this isn't currently possible.
Does Rainbird publish any documentation on API functionality? I presume they don't ;)

Thanks!

**Comment by allenporter** [comment]
Hello, perhaps you are you looking for`get_weather_and_status` and `get_schedule` methods that are documented here?
- https://allenporter.github.io/pyrainbird/pyrainbird/async_client.html#AsyncRainbirdController.get_weather_and_status 
- https://allenporter.github.io/pyrainbird/pyrainbird/async_client.html#AsyncRainbirdController.get_schedule

See https://github.com/allenporter/pyrainbird/blob/main/CONTRIBUTING.md#running-mitmproxy if you are looking to examine the API in realtime at a lower level.

**Comment by allenporter** [comment]
You can see https://github.com/allenporter/pyrainbird/blob/main/examples/rainbird_tool.py for a simple CLI with lots of things you can do (see `--help`)

**Comment by CygnusX2112** [comment]
Thanks for the quick responses. 

I wasn't looking at the data class documentation closely enough to see that these classes are nested in the values returned in **Schedule** and in  **WeatherAndStatus** ->  **Weather**.
Apologies for overlooking that.

I'm not a python native so to speak, so when using the tool script I was making some assumptions based on what values were being returned.

`get_weather_and_status FF:FF:FF:FF:FF:FF US 99999`
`WeatherAndStatus(stick_id='FF:FF:FF:FF:FF:FF', controller=Controller(available_stations=[], custom_name=None, custom_program_names={}, custom_station_names={}), forecasted_rain={}, weather=Weather(city=None, forecast=[], location=None, time_zone_id=None, time_zone_raw_offset=None))`

`get_schedule`
`Schedule(controller_info=ControllerInfo(station_delay=0, rain_delay=0, rain_sensor=False), programs=[])`

Maybe i just don't know what I'm doing? (Very likely)

**Reply by allenporter** [reply]
You have at least two problems here:
(1) Looks like you're just making up values in that first example so the rainbird api is giving you back empty results. You'll need to figure out how to get valid values to stick in there

(2) I get back valid programs from my device so either your device is not supported by this library (because it has a different format) or it doesn't support programs
```
$  examples/rainbird_tool.py get_schedule         
Schedule(controller_info=ControllerInfo(station_delay=0, rain_delay=0, rain_sensor=0), programs=[Program(program=0, frequency=<ProgramFrequency.CUSTOM: 0>, ....  <snip>
```

**Reply by allenporter** [reply]
Also see the link I mentioned above "... f you are looking to examine the API in realtime at a lower level" if you are serious about trying to do something here, then i'd look at setting up mitm proxy with the app to see whats going on for your actual device and the real app.

**Reply by CygnusX2112** [reply]
(1) No values were made up, MAC address and Zip code were redacted for obvious reasons.
(2) That is what I was expecting to see.

I'll look into the mitm proxy

**Comment by allenporter** [comment]
OK keep in mind for #1 i believe its using a cloud API.
